### PR TITLE
Bugfix/#29 - fix off method

### LINKED ISSUE #29: off should accept callback as parameter
**Describe the bug**
Currently there is no way to detach an exact listener - you need to remove all events from element

**To Reproduce**
1. Attach event two events with `.on()` on element
2. Try to remove a single one with `.off()` like `.off('click', handler)`

**Expected behavior**
Only single event is removed

### PR DESCRIPTION
**Pull request by grazik**
- add callback parameter to the `off` method
- `_e[event]` now contains array of tuples - [function name, handler]
- `off` can be called:
     - without arguments - all event handlers are removed
     - with events only - all handlers for provided events are removed
     - with events and handler - the given handler is removed from provided events

Fixes #29

### CODE REVIEW

File: src/dom/element.js [hunk 1]
```diff
@@ -585,28 +585,40 @@ class Element {
 
       node._e = node._e || {};
       node._e[event] = node._e[event] || [];
-      node._e[event].push(callback);
+      node._e[event].push([cb.name, callback]);
```
Comment (xolir) [inline src/dom/element.js]: Is there difference between cb and callback? If so can this be more explicit?

File: src/dom/element.js [hunk 2]
```diff
@@ -585,28 +585,40 @@ class Element {
...
    * @param {string} events
    */
-  off(events) {
-    if (events === undefined) {
+  off(events, callback) {
+    if (events === undefined && callback === undefined) {
       this.each(function (node) {
         for (var evt in node._e) {
           node._e[evt].forEach(function (cb) {
-            node.removeEventListener(evt, cb);
```
Comment (xolir) [inline src/dom/element.js]: Same as above

File: src/dom/element.js [hunk 3]
```diff
@@ -585,28 +585,40 @@ class Element {
...
           })
         }
+        node._e = {};
       });
     }
 
     return this.eacharg(events, function (node, event) {
-      new Element(node._e ? node._e[event] : []).each(function (cb) {
-        node.removeEventListener(event, cb);
+      new Element(node._e ? node._e[event] : []).each(function (cb, idx) {
```
Comment (xolir) [inline src/dom/element.js]: I don't know how other feels, but I think I would prefer more explicit names for both cb and idx

File: test/unit/features/dom/element.spec.js [hunk 4]
```diff
@@ -82,4 +82,24 @@ describe('Element', () => {
...
+      const method = () => 'method';
+      const spy = jasmine.createSpy();
+      const element = new Element('button');
+
+      element.on('click', spy);
+      element.on('click', method);
+      element.off('click', spy);
+      element.trigger('click')
+
+      expect(element.get(0)._e.click[0][0]).toEqual('method');
```
Comment (xolir) [inline test/unit/features/dom/element.spec.js]: Can this be somehow made more readable? At first glance it is not clear why you are getting first element from first element of click

**Comment by mateuszluczak** [comment]
Are you sure your solution works? I have no luck with my test example:

```
    const dupa = () => alert('DUPA');

    @Component('.count-control')
    class CountControl {
      init() {
        this.$element.on('click', '.control-increment', dupa);
      }

      @Evt('click', '.control-increment')
      increment() {
        this.$emit('counter.increment');
      }

      @Evt('click', '.control-decrement')
      decrement() {
        this.$element.off('click', dupa);
      }
    }
```

Also upon investigation, handlerName doesn't look very unique:

Screenshot 2019-07-15 at 09 43 30

**Comment by mateuszluczak** [comment]
Now my example, works, should one like below also work?

HTML:
```
  <div class="count-control">
    <button class="control-increment">+1</button>
    <button class="control-decrement">-1</button>
    <button class="remove">REMOVE</button>
  </div>
```

JS:
```
    @Component('.count-control')
    class CountControl {
      @Evt('click', '.control-increment')
      increment() {
        this.$emit('counter.increment');
      }

      @Evt('click', '.control-decrement')
      decrement() {
        this.$emit('counter.decrement');
      }

      @Evt('click', '.remove')
      remove() {
        this.$element.off('click', this.increment)
      }
    }
```

Hitting *REMOVE* button making *+1* to stop work

**Comment by grazik** [comment]
I talked with @xolir and @michaldziubek and I was told that if the event is given by decorator `@Evt` then we assume that it was done on purpose thus the event won't be taken off.

**Pull request merged**